doc/man/gnunet-transport-check.1
.TH GNUNET-TRANSPORT-CHECK "1" "23 Dec 2006" "GNUnet"
.SH NAME
gnunet\-transport\-check \- a tool to test a GNUnet transport service
.SH SYNOPSIS
.B gnunet\-transport\-check
[\fIOPTIONS\fR]
.SH DESCRIPTION
.PP
gnunet\-transport\-check can be used to test or profile
a GNUnet transport service.  The tool can be used to test
both the correctness of the software as well as the correctness
of the configuration.  gnunet\-transport\-check features two modes,
called loopback mode and ping mode.  In loopback mode the test is limited to testing if the
transport can be used to communicate with itself (loopback).
This mode does not include communication with other peers which
may be blocked by firewalls and other general Internet connectivity
problems.  The loopback mode is particularly useful to test
the SMTP transport service since this service is fairly hard to
configure correctly and most problems can be reveiled by just
testing the loopback.  In ping mode the tool will attempt to download
peer advertisements from the URL specified in the configuration file
and then try to contact each of the peers.  Note that it is perfectly
normal that some peers do not respond, but if no peer responds something
is likely to be wrong.  The configuration is always taken
from the configuration file.  Do not run gnunetd while running
gnunet\-transport\-check since the transport services cannot
be used by two processes at the same time.
.PP
gnunet\-transport\-check will always produce an error\-message for
the NAT transport in loopback mode.  If NAT is configured in accept\-mode (as in,
accept connections from peers using network address translation),
the check will fail with the message "could not create HELO",
which is correct since the peer itself is clearly not going to
advertise itself as a NAT.  If the peer is configured in NAT\-mode,
that is, the peer is behind a NAT box, the message will be
'could not connect'.  For NAT, both messages are NOT errors
but exactly what is supposed to happen.
.PP
Similarly, a NAT\-ed peer should typically configure the TCP transport
to use port 0 (not listen on any port).  In this case,
gnunet\-transport\-check will print 'could not create HELO' for the
TCP transport.  This is also ok.  In fact, a correctly configured
peer using NAT should give just two errors (could not connect for
tcp and could not create HELO for NAT) when tested using
gnunet\-transport\-check.  The reason is, that gnunet\-transport\-check
only tests loopback connectivity, and for a NAT\-ed peer, that just
does not apply.
.PP
Note that in ping mode the HTTP download times out after 5 minutes,
so if the list of peers is very large and not all peers can be
queried within the 5 minutes the tool may abort before trying all
peers.
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=\fIFILENAME\fR
use config file (default: /etc/gnunetd.conf)
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=\fILOGLEVEL\fR
change the loglevel. Possible values for \fILOGLEVEL\fR are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.TP
\fB\-p\fR, \fB\-\-ping\fR
use ping mode (loopback mode is default)
.TP
\fB\-r\fI COUNT \fB\-\-repeat=\fICOUNT\fR
send COUNT messages in a sequence over the same connection
.TP
\fB\-s\fI SIZE \fB\-\-size=\fISIZE\fR
test using the specified message size, default is 11
.TP
\fB\-t\fI TRANSPORT\fR, \fB\-\-transport=\fITRANSPORT\fR
run using the specified transport, if not given the transports
configured in the configuration file are used.
.TP
\fB\-u \fIUSER\fR, \fB\-\-user=USER\fR
run as user USER (and if available as group USER). Note that to use this option, you will probably have to start gnunet-transport-check as
root. It is typically better to directly start gnunet-transport-check as that user instead.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-V\fR, \fB\-\-verbose\fR
be verbose
.SH NOTES
gnunet\-transport\-check can run for a long time, depending on
how high you have set the \fICOUNT\fR level. Run first with small numbers
for \fICOUNT\fR to get an initial estimate on the runtime.
.SH FILES
.TP
/etc/gnunetd.conf
default gnunetd configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunetd.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/server/gnunet-transport-check.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2003, 2004, 2006 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file server/gnunet-transport-check.c
 * @brief Test for the transports.
 * @author Christian Grothoff
 *
 * This utility can be used to test if a transport mechanism for
 * GNUnet is properly configured.
 */

#include "platform.h"
#include "gnunet_util.h"
#include "gnunet_directories.h"
#include "gnunet_protocols.h"
#include "gnunet_transport_service.h"
#include "gnunet_identity_service.h"
#include "gnunet_pingpong_service.h"
#include "gnunet_bootstrap_service.h"
#include "core.h"
#include "connection.h"
#include "handler.h"
#include "startup.h"

#define DEBUG_TRANSPORT_CHECK GNUNET_NO

static struct GNUNET_Semaphore *sem;

static int terminate;

static unsigned long long timeout;

static GNUNET_Transport_ServiceAPI *transport;

static GNUNET_Identity_ServiceAPI *identity;

static GNUNET_Pingpong_ServiceAPI *pingpong;

static GNUNET_Bootstrap_ServiceAPI *bootstrap;

static int ok;

static int ping;

static char *expectedValue;

static unsigned long long expectedSize;

static struct GNUNET_GC_Configuration *cfg;

static struct GNUNET_GE_Context *ectx;

static struct GNUNET_CronManager *cron;

static char *cfgFilename = GNUNET_DEFAULT_DAEMON_CONFIG_FILE;

static void
semUp (void *arg)
{
  struct GNUNET_Semaphore *sem = arg;

  terminate = GNUNET_YES;
  GNUNET_semaphore_up (sem);
}

static int
noiseHandler (const GNUNET_PeerIdentity * peer,
              const GNUNET_MessageHeader * msg, GNUNET_TSession * s)
{
  if ((ntohs (msg->size) ==
       sizeof (GNUNET_MessageHeader) + expectedSize) &&
      (0 == memcmp (expectedValue, &msg[1], expectedSize)))
    ok = GNUNET_YES;
  GNUNET_semaphore_up (sem);
  return GNUNET_OK;
}

/**
 * Test the given transport API.
 */
static void
testTAPI (GNUNET_TransportAPI * tapi, void *ctx)
{
  int *res = ctx;
  GNUNET_MessageHello *helo;
  GNUNET_TSession *tsession;
  unsigned long long repeat;
  unsigned long long total;
  GNUNET_CronTime start;
  GNUNET_CronTime end;
  GNUNET_MessageHeader *noise;
  int ret;

  GNUNET_GE_ASSERT (ectx, tapi != NULL);
  if (tapi->protocol_number == GNUNET_TRANSPORT_PROTOCOL_NUMBER_NAT)
    {
      *res = GNUNET_OK;
      return;                   /* NAT cannot be tested */
    }
  helo = tapi->hello_create ();
  if (helo == NULL)
    {
      fprintf (stderr, _("`%s': Could not create hello.\n"),
               tapi->transport_name);
      *res = GNUNET_SYSERR;
      return;
    }
  tsession = NULL;
  if (GNUNET_OK != tapi->connect (helo, &tsession, GNUNET_NO))
    {
      fprintf (stderr, _("`%s': Could not connect.\n"), tapi->transport_name);
      *res = GNUNET_SYSERR;
      GNUNET_free (helo);
      return;
    }
  GNUNET_free (helo);
  if (-1 == GNUNET_GC_get_configuration_value_number (cfg,
                                                      "TRANSPORT-CHECK",
                                                      "REPEAT",
                                                      1,
                                                      (unsigned long) -1,
                                                      1, &repeat))
    {
      *res = GNUNET_SYSERR;
      return;
    }
  total = repeat;
  sem = GNUNET_semaphore_create (0);
  start = GNUNET_get_time ();
  noise = GNUNET_malloc (expectedSize + sizeof (GNUNET_MessageHeader));
  noise->type = htons (GNUNET_P2P_PROTO_NOISE);
  noise->size = htons (expectedSize + sizeof (GNUNET_MessageHeader));
  memcpy (&noise[1], expectedValue, expectedSize);
  while ((repeat > 0) && (GNUNET_shutdown_test () == GNUNET_NO))
    {
      repeat--;
      ok = GNUNET_NO;
      ret = GNUNET_NO;
      while (ret == GNUNET_NO)
        ret =
          GNUNET_CORE_connection_send_plaintext (tsession, (char *) noise,
                                                 ntohs (noise->size));
      if (ret != GNUNET_OK)
        {
          fprintf (stderr, _("`%s': Could not send.\n"),
                   tapi->transport_name);
          *res = GNUNET_SYSERR;
          tapi->disconnect (tsession);
          GNUNET_semaphore_destroy (sem);
          GNUNET_free (noise);
          return;
        }
      GNUNET_cron_add_job (cron, &semUp, timeout, 0, sem);
      GNUNET_semaphore_down (sem, GNUNET_YES);
      GNUNET_cron_suspend_jobs (cron, GNUNET_NO);
      GNUNET_cron_del_job (cron, &semUp, 0, sem);
      GNUNET_cron_resume_jobs (cron, GNUNET_NO);
      if (ok != GNUNET_YES)
        {
          FPRINTF (stderr,
                   _("`%s': Did not receive message within %llu ms.\n"),
                   tapi->transport_name, timeout);
          *res = GNUNET_SYSERR;
          tapi->disconnect (tsession);
          GNUNET_semaphore_destroy (sem);
          GNUNET_free (noise);
          return;
        }
    }
  GNUNET_free (noise);
  end = GNUNET_get_time ();
  if (GNUNET_OK != tapi->disconnect (tsession))
    {
      fprintf (stderr, _("`%s': Could not disconnect.\n"),
               tapi->transport_name);
      *res = GNUNET_SYSERR;
      GNUNET_semaphore_destroy (sem);
      return;
    }
  GNUNET_semaphore_destroy (sem);
  printf (_
          ("`%s' transport OK.  It took %ums to transmit %llu messages of %llu bytes each.\n"),
          tapi->transport_name,
          (unsigned int) ((end - start) / GNUNET_CRON_MILLISECONDS), total,
          expectedSize);
}

static void
pingCallback (void *unused)
{
  ok = GNUNET_YES;
  GNUNET_semaphore_up (sem);
}

static void
testPING (const GNUNET_MessageHello * xhello, void *arg)
{
  int *stats = arg;
  GNUNET_TSession *tsession;
  GNUNET_MessageHello *hello;
  GNUNET_MessageHello *myHello;
  GNUNET_MessageHeader *ping;
  char *msg;
  int len;
  GNUNET_PeerIdentity peer;
  unsigned long long verbose;

  stats[0]++;                   /* one more seen */
  if (GNUNET_NO == transport->test_available (ntohs (xhello->protocol)))
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_DEBUG | GNUNET_GE_REQUEST | GNUNET_GE_USER,
                     _(" Transport %d is not being tested\n"),
                     ntohs (xhello->protocol));
      return;
    }
  if (ntohs (xhello->protocol) == GNUNET_TRANSPORT_PROTOCOL_NUMBER_NAT)
    return;                     /* NAT cannot be tested */

  stats[1]++;                   /* one more with transport 'available' */
  GNUNET_GC_get_configuration_value_number (cfg,
                                            "GNUNET",
                                            "VERBOSE",
                                            0, (unsigned long long) -1, 0,
                                            &verbose);
  if (verbose > 0)
    {
      char *str;
      void *addr;
      unsigned int addr_len;
      int have_addr;

      have_addr = transport->hello_to_address (xhello, &addr, &addr_len);
      if (have_addr == GNUNET_NO)
        {
          str = GNUNET_strdup ("NAT");  /* most likely */
        }
      else
        {
          str = GNUNET_get_ip_as_string (addr, addr_len, GNUNET_YES);
          GNUNET_free (addr);
        }
      fprintf (stderr, _("\nContacting `%s'."), str);
      GNUNET_free (str);
    }
  else
    fprintf (stderr, ".");
  hello = GNUNET_malloc (ntohs (xhello->header.size));
  memcpy (hello, xhello, ntohs (xhello->header.size));

  myHello = transport->hello_create (ntohs (xhello->protocol));
  if (myHello == NULL)
    /* try NAT */
    myHello = transport->hello_create (GNUNET_TRANSPORT_PROTOCOL_NUMBER_NAT);
  if (myHello == NULL)
    {
      GNUNET_free (hello);
      return;
    }
  if (verbose > 0)
    fprintf (stderr, ".");
  tsession = NULL;
  peer = hello->senderIdentity;
  tsession = transport->connect (hello, __FILE__, GNUNET_NO);
  GNUNET_free (hello);
  if (tsession == NULL)
    {
      fprintf (stderr, _(" Connection failed\n"));
      return;
    }
  if (tsession == NULL)
    {
      GNUNET_GE_BREAK (ectx, 0);
      fprintf (stderr, _(" Connection failed (bug?)\n"));
      return;
    }
  if (verbose > 0)
    fprintf (stderr, ".");

  sem = GNUNET_semaphore_create (0);
  ping = pingpong->pingUser (&peer, &pingCallback, NULL, GNUNET_YES, rand ());
  len = ntohs (ping->size) + ntohs (myHello->header.size);
  msg = GNUNET_malloc (len);
  memcpy (msg, myHello, ntohs (myHello->header.size));
  memcpy (&msg[ntohs (myHello->header.size)], ping, ntohs (ping->size));
  GNUNET_free (myHello);
  GNUNET_free (ping);
  /* send ping */
  ok = GNUNET_NO;
  if (GNUNET_OK != GNUNET_CORE_connection_send_plaintext (tsession, msg, len))
    {
      fprintf (stderr, "Send failed.\n");
      GNUNET_free (msg);
      transport->disconnect (tsession, __FILE__);
      return;
    }
  GNUNET_free (msg);
  if (verbose > 0)
    fprintf (stderr, ".");
  /* check: received pong? */
#if DEBUG_TRANSPORT_CHECK
  GNUNET_GE_LOG (ectx, GNUNET_GE_DEBUG | GNUNET_GE_REQUEST | GNUNET_GE_USER,
                 "Waiting for PONG\n");
#endif
  terminate = GNUNET_NO;
  GNUNET_cron_add_job (cron, &semUp, timeout, 5 * GNUNET_CRON_SECONDS, sem);
  GNUNET_semaphore_down (sem, GNUNET_YES);

  if (verbose > 0)
    {
      if (ok != GNUNET_YES)
        FPRINTF (stderr, _("Timeout after %llums.\n"), timeout);
      else
        fprintf (stderr, _("OK!\n"));
    }
  GNUNET_cron_suspend_jobs (cron, GNUNET_NO);
  GNUNET_cron_del_job (cron, &semUp, 5 * GNUNET_CRON_SECONDS, sem);
  GNUNET_cron_resume_jobs (cron, GNUNET_NO);
  GNUNET_semaphore_destroy (sem);
  sem = NULL;
  transport->disconnect (tsession, __FILE__);
  if (ok == GNUNET_YES)
    stats[2]++;
}

static int
testTerminate (void *arg)
{
  if (GNUNET_shutdown_test () == GNUNET_NO)
    return GNUNET_YES;
  return GNUNET_NO;
}

/**
 * All gnunet-transport-check command line options
 */
static struct GNUNET_CommandLineOption gnunettransportcheckOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Tool to test if GNUnet transport services are operational.")),        /* -h */
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'p', "ping", NULL,
   gettext_noop ("ping peers from HOSTLISTURL that match transports"),
   0, &GNUNET_getopt_configure_set_one, &ping},
  {'r', "repeat", "COUNT",
   gettext_noop ("send COUNT messages"),
   1, &GNUNET_getopt_configure_set_option, "TRANSPORT-CHECK:REPEAT"},
  {'s', "size", "SIZE",
   gettext_noop ("send messages with SIZE bytes payload"),
   1, &GNUNET_getopt_configure_set_option, "TRANSPORT-CHECK:SIZE"},
  {'t', "transport", "TRANSPORT",
   gettext_noop ("specifies which TRANSPORT should be tested"),
   1, &GNUNET_getopt_configure_set_option, "GNUNETD:TRANSPORTS"},
  {'T', "timeout", "MS",
   gettext_noop ("specifies after how many MS to time-out"),
   1, &GNUNET_getopt_configure_set_option, "TRANSPORT-CHECK:TIMEOUT"},
  {'u', "user", "LOGIN",
   gettext_noop ("run as user LOGIN"),
   1, &GNUNET_getopt_configure_set_option, "GNUNETD:USER"},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_VERBOSE,
  {'X', "Xrepeat", "X",
   gettext_noop ("repeat each test X times"),
   1, &GNUNET_getopt_configure_set_option, "TRANSPORT-CHECK:X-REPEAT"},
  GNUNET_COMMAND_LINE_OPTION_END,
};

int
main (int argc, char *const *argv)
{
  int res;
  unsigned long long Xrepeat;
  char *trans;
  int stats[3];
  int pos;

  res = GNUNET_init (argc,
                     argv,
                     "gnunet-transport-check",
                     &cfgFilename, gnunettransportcheckOptions, &ectx, &cfg);
  if ((res == -1)
      || (GNUNET_OK != GNUNET_CORE_startup_change_user (ectx, cfg)))
    {
      GNUNET_fini (ectx, cfg);
      return -1;
    }

  if (-1 == GNUNET_GC_get_configuration_value_number (cfg,
                                                      "TRANSPORT-CHECK",
                                                      "SIZE",
                                                      1, 60000, 12,
                                                      &expectedSize))
    {
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  if (-1 == GNUNET_GC_get_configuration_value_number (cfg,
                                                      "TRANSPORT-CHECK",
                                                      "TIMEOUT",
                                                      1,
                                                      60 *
                                                      GNUNET_CRON_SECONDS,
                                                      3 * GNUNET_CRON_SECONDS,
                                                      &timeout))
    {
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  /* have no client disturb us! */
  GNUNET_GC_set_configuration_value_string (cfg,
                                            NULL,
                                            "TCPSERVER", "DISABLE", "YES");
  expectedValue = GNUNET_malloc (expectedSize);
  pos = expectedSize;
  expectedValue[--pos] = '\0';
  while (pos-- > 0)
    expectedValue[pos] = 'A' + (pos % 26);

  trans = NULL;
  if (-1 == GNUNET_GC_get_configuration_value_string (cfg,
                                                      "GNUNETD",
                                                      "TRANSPORTS",
                                                      "udp tcp http", &trans))
    {
      GNUNET_free (expectedValue);
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  GNUNET_GE_ASSERT (ectx, trans != NULL);
  if (ping)
    printf (_("Testing transport(s) %s\n"), trans);
  else
    printf (_("Available transport(s): %s\n"), trans);
  GNUNET_free (trans);
  if (!ping)
    {
      /* disable blacklists (loopback is often blacklisted)... */
      GNUNET_GC_set_configuration_value_string (cfg, ectx, "TCP",
                                                "BLACKLISTV4", "");
      GNUNET_GC_set_configuration_value_string (cfg, ectx, "TCP",
                                                "BLACKLISTV6", "");
      GNUNET_GC_set_configuration_value_string (cfg, ectx, "UDP",
                                                "BLACKLISTV4", "");
      GNUNET_GC_set_configuration_value_string (cfg, ectx, "UDP",
                                                "BLACKLISTV6", "");
      GNUNET_GC_set_configuration_value_string (cfg, ectx, "HTTP",
                                                "BLACKLISTV4", "");
      GNUNET_GC_set_configuration_value_string (cfg, ectx, "HTTP",
                                                "BLACKLISTV6", "");
    }
  cron = GNUNET_cron_create (ectx);
  if (GNUNET_OK != GNUNET_CORE_init (ectx, cfg, cron, NULL))
    {
      GNUNET_free (expectedValue);
      GNUNET_cron_destroy (cron);
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  GNUNET_CORE_connection_init (ectx, cfg, NULL, cron);
  GNUNET_CORE_plaintext_register_handler (GNUNET_P2P_PROTO_NOISE,
                                          &noiseHandler);
  GNUNET_CORE_p2p_enable_processing ();
  identity = GNUNET_CORE_request_service ("identity");
  transport = GNUNET_CORE_request_service ("transport");
  pingpong = GNUNET_CORE_request_service ("pingpong");
  GNUNET_cron_start (cron);

  GNUNET_GC_get_configuration_value_number (cfg,
                                            "TRANSPORT-CHECK",
                                            "X-REPEAT",
                                            1, (unsigned long long) -1, 1,
                                            &Xrepeat);
  res = GNUNET_OK;
  if (ping)
    {
      bootstrap = GNUNET_CORE_request_service ("bootstrap");

      stats[0] = 0;
      stats[1] = 0;
      stats[2] = 0;
      bootstrap->bootstrap (&testPING, &stats[0], &testTerminate, NULL);
      printf (_
              ("\n%d out of %d peers contacted successfully (%d times transport unavailable).\n"),
              stats[2], stats[1], stats[0] - stats[1]);
      GNUNET_CORE_release_service (bootstrap);
    }
  else
    {
      while ((Xrepeat-- > 0) && (GNUNET_shutdown_test () == GNUNET_NO))
        transport->iterate_available (&testTAPI, &res);
    }
  GNUNET_cron_stop (cron);
  GNUNET_CORE_release_service (identity);
  GNUNET_CORE_release_service (transport);
  GNUNET_CORE_release_service (pingpong);
  GNUNET_CORE_p2p_disable_processing ();
  GNUNET_CORE_plaintext_unregister_handler (GNUNET_P2P_PROTO_NOISE,
                                            &noiseHandler);
  GNUNET_CORE_connection_done ();
  GNUNET_CORE_done ();
  GNUNET_free (expectedValue);
  GNUNET_cron_destroy (cron);
  GNUNET_fini (ectx, cfg);

  if (res != GNUNET_OK)
    return -1;
  return 0;
}


/* end of gnunet-transport-check */
